src/anyio/_core/_sockets.py

```diff
@@ -184,7 +187,7 @@ async def try_connect(remote_host: str, event: Event):
     if tls or tls_hostname or ssl_context:
         try:
             return await TLSStream.wrap(connected_stream, server_side=False,
-                                        hostname=tls_hostname or remote_host,
+                                        hostname=tls_hostname or cast(Optional[str], remote_host),
```

Comment:
remote_host is a `IPAddress | None` which works at runtime because it supports `.__str__(`

TLSStream.wrap could be changed to accept an IPAddress or we could str(remote_host) here

I just did a cast because it was the quickest fix